### PR TITLE
Improve confidence-based taxonomy prediction performance

### PR DESCRIPTION
**Pull request by SirDavidLudwig**
This PR significantly improves the efficiency and performance of confidence-based taxonomy prediction by removing the need for `deepcopy`, resulting in a ~750% speed-up to sequences/second measured on our machine.

### CODE REVIEW

File: q2_feature_classifier/_skl.py
Comment (BenKaehler) [inline q2_feature_classifier/_skl.py]: Suggest changing this to
`classes = list(zip(split_classes, class_probs))`
and deleting line 83.

File: q2_feature_classifier/_skl.py
Comment (BenKaehler) [inline q2_feature_classifier/_skl.py]: This is the potential error. The error arises because sometimes taxa match at a later lever (in the label) but differ earlier. For instance the Ruminococcus genus exists in both the Lachnospiraceae and Ruminococcaceae families in Greengenes.

I suggest matching on all the levels up to the current one. Something like
```
for rank in range(1, len(taxon) + 1):
    cum_prob = sum(cls[1] for cls in classes if cls[0][:rank] == taxon[:rank])
```

File: q2_feature_classifier/_skl.py
Comment (BenKaehler) [inline q2_feature_classifier/_skl.py]: Suggest deleting this line and replacing the original definition of classes with a list rather than a zip object.

I'm fairly confident that the result would be neater, less confusing, and more efficient.

I might be wrong about the efficiency, though, and it might be faster to create the zip object at every iteration of this loop than to create the whole list before iterating. You have done performance tests so you might like to test both.

File: q2_feature_classifier/_skl.py
Comment (BenKaehler) [inline q2_feature_classifier/_skl.py]: Suggest replacing this line with
`result = taxon[:rank]`
if you make the suggested change at line 80.

**Review by BenKaehler — CHANGES_REQUESTED** [review]
Hi @SirDavidLudwig, thanks very much for this PR. That improvement in efficiency is great.

There is one potential error (although it doesn't show up in the unit tests) and I have one suggestion. Please refer to the comments that I attached to the code.

**Comment by SirDavidLudwig** [comment]
If we can safely make the assumption that the model's classes are ordered alphabetically (which appears to be the case, but I could be wrong), I'd like to propose a new tree based solution to this problem. Not only does it address the issue you mentioned, but the search space is heavily optimized, allowing for a 100X speed-up. I attached a simple bar plot comparing some numbers that I observed in my testing. If this assumption is safe to make, I'd like to update this PR to reflect these new changes.

image

**Comment by SirDavidLudwig** [comment]
I didn't realize Github automatically updates pull requests when code is committed. I can submit this idea as a different pull request and revert back to the original changes we discussed.

File: q2_feature_classifier/_skl.py
Comment (SirDavidLudwig) [reply q2_feature_classifier/_skl.py]: Hmm... This makes sense, but the redundant checking will also have a significant performance impact. We could go with this for now though, but I believe there's a better solution that could be implemented.

File: q2_feature_classifier/_skl.py
Comment (SirDavidLudwig) [reply q2_feature_classifier/_skl.py]: The change in efficiency would be negligible here... But I agree, it would be cleaner to create a list instance once in the beginning of the loop.

File: q2_feature_classifier/_skl.py
Comment (BenKaehler) [inline q2_feature_classifier/_skl.py]: I think the linter is failing somewhere around here. Please fix.

File: q2_feature_classifier/_skl.py
Comment (BenKaehler) [inline q2_feature_classifier/_skl.py]: Please pass the `separator` that is passed to `predict` down to this level, so that it is guaranteed to be consistent.

File: q2_feature_classifier/_skl.py
Comment (BenKaehler) [inline q2_feature_classifier/_skl.py]: As you state in the comments, you assume that `pipeline.classes_` is sorted. The sklearn classes I've looked at do sort their classes, but please add an assertion here to check that it is not violated in the future.

**Review by BenKaehler — CHANGES_REQUESTED** [review]
Thanks @SirDavidLudwig, using the natural order of the classes in the probability vector is a clever idea. That should also fix the issue I raised with the last version of this PR.

Please address the minor comments that I've raised and I will approve.

**Comment by BenKaehler** [comment]
Hi @SirDavidLudwig, just an afterthought: could you please also add a brief "here be dragons"-style comment to the new `_TaxonPredicitonNode` class to explain to future maintainers why it works?

File: q2_feature_classifier/_skl.py
Comment (SirDavidLudwig) [reply q2_feature_classifier/_skl.py]: Since the tree structure requires the classes to be sorted, I put the assertion inside the `create_tree` method.

File: q2_feature_classifier/_skl.py
Comment (BenKaehler) [inline q2_feature_classifier/_skl.py]: Sorry @SirDavidLudwig, did you mean k__Bacteria here?

**Review by BenKaehler — CHANGES_REQUESTED** [review]
Thanks @SirDavidLudwig, your changes look great. Just one minor typo in your comment that I think needs fixing. Thanks again.

File: q2_feature_classifier/_skl.py
Comment (SirDavidLudwig) [reply q2_feature_classifier/_skl.py]: Yes, sorry about that. Just pushed the fix.

**Comment by BenKaehler** [comment]
Hi @ebolyen, are you ok with me merging this, or should it get another review? Thanks!

**Review by misialq — CHANGES_REQUESTED** [review]
Hey @SirDavidLudwig, thanks for your contribution! It would be super lovely if you could ad a couple of unit tests specific to your new data class, demonstrating the usage of the class methods/properties and the edge cases they are trying to account for. Thanks!

**Comment by BenKaehler** [comment]
Sorry for the delay @SirDavidLudwig!

After being unable to contact Evan, I decided to do a couple more tests.

I ran the Silva 138.1 515f-806r-animal-distal-gut-classifier off Zenodo on 2,689 sequences from the PD Mouse dataset with 11 cores on my aging laptop.

It took 1m47s using the classifier proposed here and 6m14s using the existing classifier, so that's a substantial improvement, even with the unfairness of parallel computing and the latency of loading a large classifier 11 times.

All taxonomic classifications were identical and the largest difference in confidence estimates was 9.6e-14.

Thanks very much @misialq for running another set of eyes over the changes. You are, of course, correct that more unit tests should be added. I think once those are in we're definitely ready to merge.

These are my suggestions for some unit tests that could be added to the end of `ClassifierTests` in `test_classifier.py` (thanks copilot):

``` python
    def test_TaxonNode_create_tree(self):
        classes = ['a;b;c', 'a;b;d', 'a;e;f', 'a;e;g']
        separator = ';'
        tree = _TaxonNode.create_tree(classes, separator)
        self.assertEqual(
            tree.children['a'].children['b'].children['c'].name, 'c')
        self.assertEqual(
            tree.children['a'].children['b'].children['d'].name, 'd')
        self.assertEqual(
            tree.children['a'].children['e'].children['f'].name, 'f')
        self.assertEqual(
            tree.children['a'].children['e'].children['g'].name, 'g')
    
    def test_TaxonNode_range(self):
        classes = ['a;b;c', 'a;b;d', 'a;e;f', 'a;e;g']
        separator = ';'
        tree = _TaxonNode.create_tree(classes, separator)
        self.assertEqual(
            tree.children['a'].children['b'].children['c'].range, range(0, 1))
        self.assertEqual(
            tree.children['a'].children['b'].children['d'].range, range(1, 2))
        self.assertEqual(
            tree.children['a'].children['e'].children['f'].range, range(2, 3))
        self.assertEqual(
            tree.children['a'].children['e'].children['g'].range, range(3, 4))
        self.assertEqual(
            tree.children['a'].children['b'].range, range(0, 2))
        self.assertEqual(
            tree.children['a'].children['e'].range, range(2, 4))
    
    def test_TaxonNode_num_leaf_nodes(self):
        classes = ['a;b;c', 'a;b;d', 'a;e;f', 'a;e;g']
        separator = ';'
        tree = _TaxonNode.create_tree(classes, separator)
        self.assertEqual(tree.num_leaf_nodes, 4)
        self.assertEqual(tree.children['a'].num_leaf_nodes, 4)
        self.assertEqual(tree.children['a'].children['b'].num_leaf_nodes, 2)
        self.assertEqual(tree.children['a'].children['e'].num_leaf_nodes, 2)
```

**Comment by SirDavidLudwig** [comment]
I added the above unit tests to `ClassifierTests` as suggested. Let me know if there's anything else. Thanks!

**Review by BenKaehler — APPROVED** [review]
Perfect. Thanks @SirDavidLudwig!

**Pull request merged**
